File: Chapter 2/2.2/uva/12291.cpp
#include<iostream>
#include<vector>

using namespace std;

int n {}, m{};

bool inGrid(int row, int col){
    return row >= 0 && row < n && col >= 0 && col < n;
}

bool isValid(int grid[10][10], int subGrid[10][10]){
    //get top left of both
    pair<int,int> gridTL {};
    pair<int,int> subGridTL {};
    vector<pair<int,int>> markers{};

    bool isFound {false};
    for(int row=0;!isFound && row<n;row++){
        for(int col=0;!isFound && col<n;col++){
            if(grid[row][col] == 1){
                isFound = true;
                gridTL.first = row;
                gridTL.second = col;
            }
        }
    }
    if(!isFound){
        return false;
    }

    //get the subGrid vector of changes
    isFound = false;
    for(int row=0;row<m;row++){
        for(int col=0;col<m;col++){
            if(subGrid[row][col] == 1){
                if(!isFound){
                    isFound = true;
                    subGridTL.first = row;
                    subGridTL.second = col;
                }

                pair<int,int> temp {};
                temp.first = row - subGridTL.first;
                temp.second = col - subGridTL.second;
                markers.push_back(temp);
            }
        }
    }
    if(!isFound){
        return false;
    }

    //first removal of original layout
    for(auto pr:markers){
        int tempRow = gridTL.first + pr.first;
        int tempCol = gridTL.second + pr.second;

        if((!inGrid(tempRow,tempCol)) || (grid[tempRow][tempCol] == 0)){
            return false;
        } else {
            grid[tempRow][tempCol] = 0;
        }
    }

    isFound=false;
    for(int row=gridTL.first;!isFound && row<n;row++){
        for(int col=0;!isFound && col<n;col++){
            if(grid[row][col] == 1){
                isFound = true;
                gridTL.first = row;
                gridTL.second = col;
            }
        }
    }
    if(!isFound){
        return false;
    }

    //second removal
    for(auto pr:markers){
        int tempRow = gridTL.first + pr.first;
        int tempCol = gridTL.second + pr.second;

        if((!inGrid(tempRow,tempCol)) || (grid[tempRow][tempCol] == 0)){
            return false;
        } else {
            grid[tempRow][tempCol] = 0;
        }
    }

    //making sure nothing got left behind
    for(int row=0;row<n;row++){
        for(int col=0;col<n;col++){
            if(grid[row][col] == 1){
                return false;
            }
        }
    }

    return true;
}

int main(){
    string s;
    while((cin>>n>>m)){
        if(m==0 & n==0){
            break;
        }
        auto grid = new int[10][10];
        auto subGrid = new int[10][10];

        for(int row=0;row<n;row++){
            cin >> s;
            for(int col=0;col<n;col++){
                if(s.at(col) == '*'){
                    grid[row][col]=1;
                } else {
                    grid[row][col]=0;
                }
            }
        }

        for(int row=0;row<m;row++){
            cin >> s;
            for(int col=0;col<m;col++){
                if(s.at(col) == '*'){
                    subGrid[row][col]=1;
                } else {
                    subGrid[row][col]=0;
                }
            }
        }

        if(isValid(grid, subGrid)){
            cout << 1;
        } else {
            cout << 0;
        }
        cout << '\n';

        delete grid;
        delete subGrid;
    }
    return 0;
}

File: Chapter 2/2.2/uva/12291.py

n = 0
m = 0
grid = []
subGrid = []

def getTopLeft(grid:list) -> tuple:
    for i,row in enumerate(grid):
        for j,col in enumerate(row):
            if col == "*":
                return (i,j,True)
    return (-1,-1,False)

def inGrid(row:int, col:int):
    global n
    return 0 <= row < n and 0 <= col < n

def isValid() -> bool :
    global n,m,grid,subGrid

    markers = []
    isFound = False
    subGridTL = (-1,-1,False)
    for i,row in enumerate(subGrid):
        for j,col in enumerate(row):
            if col == "*":
                if not isFound:
                    isFound = True
                    subGridTL = (i,j,True)
                
                markers.append((i-subGridTL[0], j-subGridTL[1]))
    
    if not subGridTL[2]:
        return False

    gridTL = getTopLeft(grid)
    if not gridTL[2]:
        return False
    
    #first removal
    for marker in markers:
        tempRow = gridTL[0] + marker[0]
        tempCol = gridTL[1] + marker[1]

        if not inGrid(tempRow,tempCol):
            return False

        if grid[tempRow][tempCol] != "*":
            return False

        grid[tempRow][tempCol] = "." 

    gridTL = getTopLeft(grid)
    if not gridTL[2]:
        return False
    
    #second removal
    for marker in markers:
        tempRow = gridTL[0] + marker[0]
        tempCol = gridTL[1] + marker[1]

        if not inGrid(tempRow,tempCol):
            return False

        if grid[tempRow][tempCol] != "*":
            return False

        grid[tempRow][tempCol] = "." 

    #making sure nothing was left
    for row in grid:
        for col in row:
            if col == "*":
                return False

    return True

while True:
    n, m = map(int,input().split(" "))
    if n==0 and m==0:
        break

    grid = []
    subGrid = []

    for row in range(n):
        grid.append(list(input()))
    
    for row in range(m):
        subGrid.append(list(input()))

    if isValid():
        print(1)
    else:
        print(0)